src/ui/components/Overlay/index.js

```diff
@@ -1,80 +1,110 @@
+  }
+  _this.hide();
+};
+
+type Props = {|
+  className?: string,
+  children: React.Element<any>,
+  id: string,
+  onEscapeOverlay?: Function,
+  visibleOnLoad?: boolean,
+  _onClickBackground: Function,
+|};
 
+type UIStateType = {|
+  visible: boolean,
+|};
+
+const initialUIState: UIStateType = { visible: false };
+
+type InternalProps = {|
+  ...Props,
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+|};
+
+export class OverlayBase extends React.Component<InternalProps> {
   static defaultProps = {
     visibleOnLoad: false,
+    _onClickBackground: onClickBackground,
   };
 
-  constructor(props) {
-    super(props);
-    this.state = { visible: props.visibleOnLoad };
-  }
+  componentWillReceiveProps(nextProps: InternalProps) {
+    const { uiState: uiStateOld } = this.props;
+    const { visibleOnLoad: visibleOnLoadNew } = nextProps;
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.visibleOnLoad !== undefined) {
-      this.setState({ visible: nextProps.visibleOnLoad });
+    if (
+      (visibleOnLoadNew && !uiStateOld.visible) ||
+      uiStateOld.visible !== nextProps.uiState.visible
+    ) {
+      this.props.setUIState({ visible: visibleOnLoadNew });
     }
   }
 
-  onClickBackground = () => {
-    if (this.props.onEscapeOverlay) {
-      this.props.onEscapeOverlay();
-    }
-    this.hide();
-  };
-
   hide() {
-    this.setState({ visible: false });
+    this.props.setUIState({ visible: false });
   }
 
   show() {
-    this.setState({ visible: true });
+    this.props.setUIState({ visible: true });
   }
 
   toggle() {
-    this.setState({ visible: !this.state.visible });
+    this.props.setUIState({ visible: !this.props.uiState.visible });
   }
 
   render() {
-    const { children, className } = this.props;
+    const { children, className, id, uiState } = this.props;
+
+    invariant(children, 'The children property is required');
+    invariant(id, 'The id property is required');
 
     return (
       <div
         className={makeClassName('Overlay', className, {
-          'Overlay--visible': this.state.visible,
+          'Overlay--visible': uiState.visible,
         })}
-        ref={(ref) => {
-          this.overlayContainer = ref;
-        }}
       >
         <div
-          onClick={this.onClickBackground}
-          ref={(ref) => {
-            this.overlayBackground = ref;
-          }}
           className="Overlay-background"
+          onClick={(e) => this.props._onClickBackground(e, this)}
           role="presentation"
         />
-        <div
-          className="Overlay-contents"
-          ref={(ref) => {
-            this.overlayContents = ref;
-          }}
-        >
-          {children}
-        </div>
+        <div className="Overlay-contents">{children}</div>
       </div>
     );
   }
 }
+
+export const extractId = (ownProps: InternalProps) => {
+  return ownProps.id;
+};
+
+const Overlay: React.ComponentType<InternalProps> = compose(
```

Comment:
The type hint should be `React.ComponentType<Props>` to guarantee that callers pass in all required properties of `Props`, not `InternalProps`.

Reply:
okay, I was looking at other components (as a guideline) and saw this in many places. I saw that Internal Props uses Props so I thought this made sense to me. I can def update I just wanted to let you know where I was coming from : )

Reply:
oh! I see... this is just at the end of the file...